### PR TITLE
Make workers register to all masters

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Make workers register to all masters
1. Enabled grpc block master service on standby masters
2. Introduce SpecificMasterBlockSync on workers to heartbeat all masters and BlockSyncMaster for its abstraction
3. Introduce add worker id rpc to reach consensus on worker ids across masters
4. Journal the block location change
5. Update the heartbeat report generation and make it more fault tolerant
6. All a basic worker registration integration test 

### Why are the changes needed?

To accelerate the master failover process

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

File: core/server/master/src/main/java/alluxio/master/block/BlockMaster.java [hunk 1]
```diff
@@ -392,4 +392,11 @@ void workerRegisterStream(
    * @param workerId the workerId of target worker
    */
   void removeDecommissionedWorker(long workerId) throws NotFoundException;
+
```
Comment (elega) [inline core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:395]: this add worker id is used to keep the block id map across masters consistent so that when failover happens, the new elected primary master can still generate unique worker ids.

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 2]
```diff
@@ -920,9 +961,23 @@ public void commitBlock(long workerId, long usedBytesOnTier, String tierAlias,
                   block.get().getLength(), length);
             } else {
               mBlockMetaStore.putBlock(blockId, BlockMeta.newBuilder().setLength(length).build());
-              BlockInfoEntry blockInfo =
-                  BlockInfoEntry.newBuilder().setBlockId(blockId).setLength(length).build();
-              journalContext.append(JournalEntry.newBuilder().setBlockInfo(blockInfo).build());
+              BlockInfoEntry.Builder blockInfoBuilder =
```
Comment (elega) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:964]: journal the block location info here and replicate them to standby masters

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 3]
```diff
@@ -1091,15 +1150,59 @@ public long getWorkerId(WorkerNetAddress workerNetAddress) {
...
+    if (!mTempWorkers.add(new MasterWorkerInfo(workerId, workerNetAddress))) {
+      throw new RuntimeException("Duplicated worker ID for " + workerId + ": " + workerNetAddress);
+    }
+  }
+
+  // TODO(elega) we change the way worker id generates and makes worker ids
+  // change when we toggle the workers registering to all masters feature flag,
+  // and might cause backward compatibility issues. Need to fix this later.
+  @VisibleForTesting
+  synchronized long generateWorkerId(WorkerNetAddress addr, boolean workerRegisterToAllMasters) {
```
Comment (elega) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1186]: we changed the worker id generation algorithm to make it address deterministic

Comment (elega) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1186]: actually do we have to make the change? can we journal the worker id instead?

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 4]
```diff
@@ -1310,6 +1413,35 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
     // by the LostWorkerDetectionHeartbeatExecutor
     worker.updateLastUpdatedTimeMs();
 
+    if (mWorkerRegisterToAllMasters && mPrimarySelector.getState() == NodeState.STANDBY) {
```
Comment (elega) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1416]: on standby master, heartbeat can come and be processed from a journal.
e.g. 
A journal that creates a block is processed after a worker master heartbeat to remove such block.
This will result in some state inconsistency so we wait here for a best effort first before we ignore the block change in heartbeat

File: core/server/worker/src/main/java/alluxio/worker/block/AllMasterRegistrationBlockWorker.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      ProcessUtils.fatalError(LOG, e, "Failed to register with primary master");
+    }
+    LOG.info("The worker has registered with primary master, address {}",
+        mFileSystemMasterClient.getRemoteSockAddress());
+  }
+
+  @Override
+  public void start(WorkerNetAddress address) throws IOException {
+    super.start(address);
+    // Because standby masters do not take read requests,
```
Comment (elega) [inline core/server/worker/src/main/java/alluxio/worker/block/AllMasterRegistrationBlockWorker.java:135]: ^^ this

File: core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java [hunk 6]
```diff
@@ -62,12 +70,11 @@ public BlockHeartbeatReporter() {
...
 
   /**
-   * Generates the report of the block store delta in the last heartbeat period. Calling this method
-   * marks the end of a period and the start of a new heartbeat period.
+   * Generates the report of the report and clear the states.
    *
    * @return the block store delta report for the last heartbeat period
    */
-  public BlockHeartbeatReport generateReport() {
+  public BlockHeartbeatReport generateReportAndClear() {
```
Comment (elega) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:77]: this change is made to fix a existing bug in heartbeat threads.
if the report it generated but the heartbeat rpc fails, the block location info generated by the report will just be gone. so we added a revert method here as well as rename this function to generateReportAndClear to make it closer to its semantics.

File: core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java [hunk 7]
```diff
@@ -134,6 +190,7 @@ public void onStorageLost(String tierAlias, String dirPath) {
   }
 
   private void removeBlockInternal(long blockId) {
+//    System.out.println("removeBlockInternal " + blockId + " " + this);
```
Comment (elega) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:193]: will remove before merge the pr

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+    mLastSuccessfulHeartbeatMs = CommonUtils.getCurrentMs();
+  }
+
+  private RetryPolicy createEndlessRetry() {
+    return new ExponentialBackoffRetry(
+        1000, 60 * 1000, Integer.MAX_VALUE);
+  }
+
+  @Override
+  public synchronized void heartbeat() throws InterruptedException {
```
Comment (elega) [inline core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:160]: endless retries are used in worker registration and heartbeats as we don't require heartbeats to all masters to succeed

**Comment by LuQQiu** [comment]
@elega thanks for contributing!

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 9]
```diff
@@ -1076,6 +1134,10 @@ protected MasterWorkerInfo recordWorkerRegistration(long workerId) {
 
   @Override
   public long getWorkerId(WorkerNetAddress workerNetAddress) {
+    if (mStandbyMasterRpcEnabled && mPrimarySelector.getState() == NodeState.STANDBY) {
```
Comment (maobaolong) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1137]: Is this `mStandbyMasterRpcEnabled` condition check redundant?

File: core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java [hunk 10]
```diff
@@ -79,6 +86,55 @@ public BlockHeartbeatReport generateReport() {
...
+      mLostStorage.putAll(previousReport.getLostStorage());
+      mRemovedBlocks.addAll(previousReport.getRemovedBlocks());
+    }
+  }
+
+  @Override
+  public void onCommitBlockToMaster(long blockId, BlockStoreLocation location) {
+    if (mWorkerRegisterToAllMasters) {
+      synchronized (mLock) {
+        addBlockToAddedBlocks(blockId, location);
```
Comment (maobaolong) [inline core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:133]: It seems useless

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java (hunk 9 above)
Comment (elega) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1137]: yes. i added this because a bunch of tests use a default STANDBY primary selector. If i don't add this, a lot of tests will break

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 11]
```diff
@@ -920,9 +964,23 @@ public void commitBlock(long workerId, long usedBytesOnTier, String tierAlias,
...
-                  BlockInfoEntry.newBuilder().setBlockId(blockId).setLength(length).build();
-              journalContext.append(JournalEntry.newBuilder().setBlockInfo(blockInfo).build());
+              BlockInfoEntry.Builder blockInfoBuilder =
+                  BlockInfoEntry.newBuilder().setBlockId(blockId).setLength(length);
+              if (mWorkerRegisterToAllMasters) {
+                blockInfoBuilder
+                    .setBlockId(blockId)
+                    .setLength(length)
+                    .setBlockLocation(
+                        alluxio.grpc.BlockLocation.newBuilder()
```
Comment (maobaolong) [inline core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:974]: We can use a long type locationId to identify the blockLocation, so that we can reduce the transfer and storage bytes.

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+      // Unknown request
+      case Unknown:
+        LOG.error("Master heartbeat sends unknown command {}", cmd);
+        break;
+      default:
+        throw new RuntimeException("Un-recognized command from master " + cmd);
+    }
+  }
+
+  enum WorkerMasterState {
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:266]: let's make it a separate class, i feel it's more generic than it is right now

File: core/server/worker/src/test/java/alluxio/worker/block/AllMasterRegistrationBlockWorkerTest.java [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+          BlockMasterClient create(InetSocketAddress address) {
+            return mBlockMasterClient;
+          }
+        });
+  }
+
+  @Test
+  public void workerMasterRegistrationFailed() throws IOException {
+    doThrow(new RuntimeException("error")).when(mBlockMasterClient).registerWithStream(
+        anyLong(), any(), any(), any(), any(), any(), any());
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/AllMasterRegistrationBlockWorkerTest.java:64]: nit: we prefer manually extended children class to mocks, but here the tests is small enough that we accept a quick mock

File: core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java [hunk 14]
```diff
@@ -95,27 +102,27 @@ public void generateReportMove() {
...
   public void generateReportStateClear() {
     Long block1 = 1L;
     moveBlock(block1, MEM_LOC);
 
     // First report should have updates
-    BlockHeartbeatReport report = mReporter.generateReport();
+    BlockHeartbeatReport report = mReporter.generateReportAndClear();
     assertFalse(report.getAddedBlocks().isEmpty());
 
     // Second report should not have updates
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java:117]: No change required here.

Actually we prefer the master-side to be idempotent, and in the near future we should allow the worker to send dup updates. We might even prefer the worker to do so, because losing updates may result in mysterious bugs. So i'm okay if we later break this test and remove it.

File: core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java [hunk 15]
```diff
@@ -150,12 +157,49 @@ public void generateReportMoveThenRemove() {
...
+  @Test
+  public void generateUpdateThenRevert() {
+    mReporter.onCommitBlockToMaster(1, MEM_LOC);
+    mReporter.onCommitBlockToMaster(2, MEM_LOC);
+    mReporter.onCommitBlockToMaster(3, SSD_LOC);
+    mReporter.onRemoveBlockByClient(4);
+    BlockHeartbeatReport originalReport = mReporter.generateReportAndClear();
+    mReporter.onRemoveBlockByClient(1);
+    mReporter.onRemoveBlockByClient(3);
+    mReporter.onRemoveBlockByClient(5);
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java:193]: need coverage on added blocks, because thats a map<location, list> so the logic may be more complicated

File: minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A local Alluxio cluster with multiple masters using embedded journal.
+ */
+@NotThreadSafe
+public final class MultiMasterEmbeddedJournalLocalAlluxioCluster
```
Comment (jiacheliu3) [inline minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java:47]: move this to a separate PR? WDYT?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
added a bunch of comments, PTAL thanks!

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 17]
```diff
@@ -4525,6 +4525,27 @@ public String toString() {
...
+          .setDefaultValue(false)
+          .setDescription("If enabled, workers will register themselves to all masters, "
+              + "instead of primary master only. This can be used to save the "
+              + "master failover time because workers don't need to report the block locations "
+              + " to the new primary master. Can only be enabled when "
+              + Name.STANDBY_MASTER_GRPC_ENABLED + " is turned on.")
+          .setConsistencyCheckLevel(ConsistencyCheckLevel.ENFORCE)
+          .setScope(Scope.ALL)
+          .build();
+  public static final PropertyKey WORKER_ALL_MASTER_REGISTRATION_TIMEOUT_MS =
```
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:4539]: true that we should use the existing properties.

File: core/common/src/main/java/alluxio/metrics/MetricKey.java [hunk 18]
```diff
@@ -2062,6 +2062,11 @@ public static String getSyncMetricName(long mountId) {
               + "Use this metric to monitor the RPC pressure on worker.")
           .setMetricType(MetricType.GAUGE)
           .build();
+  public static final MetricKey WORKER_MASTER_REGISTRATION_SUCCESS_COUNT =
```
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:2065]: the former one

File: core/server/master/src/main/java/alluxio/master/block/BlockMaster.java [hunk 19]
```diff
@@ -392,4 +392,11 @@ void workerRegisterStream(
...
+  /**
+   * Adds the worker id to a master. Both primary and standby master implement this endpoint.
+   * @param workerId the worker id
+   * @param workerNetAddress the worker address
+   */
+  void addWorkerId(long workerId, WorkerNetAddress workerNetAddress);
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:401]: done

File: core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java [hunk 20]
```diff
@@ -1347,6 +1398,40 @@ public Command workerHeartbeat(long workerId, Map<String, Long> capacityBytesOnT
...
+      StringBuilder sb = new StringBuilder();
+      sb.append("[");
+      for (long blockIdToWait : blockIdsToWait) {
+        if (!mBlockMetaStore.getBlock(blockIdToWait).isPresent()) {
+          sb.append(blockIdToWait);
+          sb.append(" ,");
+        }
+      }
+      sb.append("]");
+      LOG.warn("Adding block ids {} for worker {} but these blocks don't exist. "
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:1430]: i only printed out the non-existing blocks here 
`        if (!mBlockMetaStore.getBlock(blockIdToWait).isPresent()) {
          sb.append(blockIdToWait);
          sb.append(" ,");
        }`

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+            mWorkerId.get(), report,
+            mBlockWorker.getStoreMeta(), this::handleMasterCommand);
+      } catch (Exception e) {
+        LOG.error("Failed to receive master heartbeat command. worker id {}", mWorkerId, e);
+      }
+      if (success) {
+        mLastSuccessfulHeartbeatMs = CommonUtils.getCurrentMs();
+        break;
+      } else {
+        mBlockHeartbeatReporter.revert(report);
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:205]: yeah i renamed this to mergeBack()

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 22]
```diff
@@ -330,7 +336,7 @@ public BlockWriter createBlockWriter(long sessionId, long blockId)
 
   @Override
   public BlockHeartbeatReport getReport() {
-    return mHeartbeatReporter.generateReport();
+    return mHeartbeatReporter.generateReportAndClear();
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:339]: sure but here isn't the best place. I added it to SpecificMasterBlockSync.java

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 23]
```diff
@@ -0,0 +1,279 @@
...
+  /**
+   * The worker registration state.
+   * If the state is NOT_REGISTERED, heartbeat will trigger a registration.
+   * During the registration process, the state will be set to REGISTERING.
+   * When the registration is done, the state will be set to REGISTERED.
+   * When the sync receives a registration command from the master during the heartbeat,
+   * the state will be reset to NOT_REGISTERED and the sync will attempt to register it again
+   * in the next heartbeat.
+   */
+  private volatile WorkerMasterState mWorkerState = WorkerMasterState.NOT_REGISTERED;
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:65]: yup let's address this in a different pr

File: core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java (hunk 14 above)
Comment (elega) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java:117]: we changed nothing... we just renamed the method generateReport() to generateReportAndClear()

File: core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java (hunk 15 above)
Comment (elega) [reply core/server/worker/src/test/java/alluxio/worker/block/BlockHeartbeatReporterTest.java:193]: yeah i updated the test

File: minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java (hunk 16 above)
Comment (elega) [reply minicluster/src/main/java/alluxio/master/MultiMasterEmbeddedJournalLocalAlluxioCluster.java:47]: sure

File: core/common/src/main/java/alluxio/worker/block/BlockHeartbeatReport.java [hunk 24]
```diff
@@ -74,4 +74,15 @@ public List<Long> getRemovedBlocks() {
   public Map<String, List<String>> getLostStorage() {
     return Collections.unmodifiableMap(mLostStorage);
   }
+
+  /**
+   * @return the number of blocks in the report
+   */
+  public int getBlockCount() {
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/worker/block/BlockHeartbeatReport.java:81]: nit: `getBlockChangeCount()` or `getBlockChangeCountEstimation()`

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 25]
```diff
@@ -4557,7 +4557,17 @@ public String toString() {
           .setConsistencyCheckLevel(ConsistencyCheckLevel.WARN)
           .setScope(Scope.WORKER)
           .build();
-
+  public static final PropertyKey WORKER_BLOCK_HEARTBEAT_REPORT_CAPACITY_THRESHOLD =
```
Comment (jiacheliu3) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:4560]: i feel `capacity` and `threshold` kinda mean the same thing. So maybe `worker.block.heartbeat.threshold`? The word "threshold" implies something will happen when this size is reached.

File: core/common/src/main/java/alluxio/metrics/MetricKey.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:2065]: what does this count mean? Like if you see the count is 100, it could mean the worker retried connecting to one particular master, or the worker restarted a few times and connected to all masters. I just feel we could probably discuss if there are certain events you'd like to capture, then we update the metric accordingly?

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 25 above)
Comment (elega) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:4560]: i'd name it to worker.block.heartbeat.report.size.threshold

File: core/common/src/main/java/alluxio/metrics/MetricKey.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:2065]: yeah but you can tell how many registration happen in a given time window which gives you some insights. e.g. if the cluster have been running for a while and there's tons of registration events, it might indicate something funky happens

[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:2065]: if you want we can split this metric to registration count plus re-registration count if you see that will be more helpful

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 26]
```diff
@@ -0,0 +1,288 @@
...
+        "Failed to register with master %s", mMasterAddress);
+  }
+
+  protected void registerWithMasterInternal()
+      throws IOException, FailedToAcquireRegisterLeaseException {
+    // The target master is not necessarily the one that allocated the workerID
+    LOG.info("Notify the master {} about the workerID {}", mMasterAddress, mWorkerId);
+    mMasterClient.notifyWorkerId(mWorkerId.get(), mWorkerAddress);
+    // TODO(elega) If worker registration to all masters happens at the same time,
+    // this might cause OOM issue. Consider limiting the worker registration concurrency.
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:156]: OOM on master or worker? The worker reg concurrency is by default controlled in https://docs.alluxio.io/os/user/stable/en/administration/Scalability-Tuning.html#worker-registration

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+    return new ExponentialBackoffRetry(
+        1000, 60 * 1000, Integer.MAX_VALUE);
+  }
+
+  @Override
+  public synchronized void heartbeat() throws InterruptedException {
+    if (mWorkerState == WorkerMasterRegistrationState.NOT_REGISTERED) {
+      // Not registered because:
+      // 1. The worker just started, we kick off the 1st registration here.
+      // 2. Master sends a registration command during
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:186]: 3. the heartbeat message becomes too big that we decide to fall back to a full re-register

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 28]
```diff
@@ -0,0 +1,288 @@
...
+    if (cmd == null) {
+      return;
+    }
+    switch (cmd.getCommandType()) {
+      // Currently unused
+      case Delete:
+        break;
+      // Master requests blocks to be removed from Alluxio managed space.
+      case Free:
+        mAsyncBlockRemover.addBlocksToDelete(cmd.getDataList());
```
Comment (jiacheliu3) [inline core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:263]: interestingly, double check what we do if the master is standby?

File: tests/src/test/java/alluxio/server/worker/WorkerAllMasterRegistrationTest.java [hunk 29]
```diff
@@ -0,0 +1,361 @@
...
+    fis = mCluster.getClient().openFile(fileUri);
+    assertEquals(fileContent, IOUtils.toString(fis, Charset.defaultCharset()));
+  }
+
+  /**
+   * Tests the worker can re-register with masters if its heartbeat failed too many times.
+   * The registration is used to address OOM issue.
+   */
+  @Test
+  public void blockHeartbeatReportOOM() throws Exception {
```
Comment (jiacheliu3) [inline tests/src/test/java/alluxio/server/worker/WorkerAllMasterRegistrationTest.java:280]:
```suggestion
  public void heartbeatFallsBackToRegister() throws Exception {
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 30]
```diff
@@ -4546,7 +4557,17 @@ public String toString() {
...
+      intBuilder(Name.WORKER_BLOCK_HEARTBEAT_REPORT_CAPACITY_THRESHOLD)
+          .setDefaultValue(1_000_000)
+          .setDescription("To avoid OOM issue, in all master registration mode, "
+              + "after a failed worker-master block sync,"
+              + "if the block heartbeat report capacity exceeds a threshold, "
+              + "the worker will clear the reporter and set the worker state unregistered, "
+              + "so that the heartbeat thread can trigger a registration again.")
```
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:4567]: sure i adopted your desc

File: core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java [hunk 31]
```diff
@@ -79,6 +86,73 @@ public BlockHeartbeatReport generateReport() {
...
+          if (removedBlocksSet.contains(blockId)) {
+            needToRemoveBlock = true;
+            break;
+          }
+        }
+        final List<Long> blockIdsToAdd;
+        if (!needToRemoveBlock) {
+          blockIdsToAdd = blockIds;
+        } else {
+          blockIdsToAdd = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:125]: how? essentially you are removing some elements from an array list, which is apparently an inefficient operation

[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/server/worker/src/main/java/alluxio/worker/block/BlockHeartbeatReporter.java:125]: i forgot it's a list...

File: core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java [hunk 32]
```diff
@@ -0,0 +1,179 @@
...
+  /**
+   * Creates a block sync master group.
+   * @param masterAddresses the master addresses to sync
+   * @param blockWorker the block worker instance
+   */
+  public BlockSyncMasterGroup(
+      List<InetSocketAddress> masterAddresses,
+      BlockWorker blockWorker
+  ) throws IOException {
+    // TODO(jiacheng/elega): handle master quorum changes
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java:64]: https://github.com/Alluxio/alluxio/issues/16898 
updated the comment

File: core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java [hunk 33]
```diff
@@ -0,0 +1,179 @@
...
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * An abstraction layer that manages the worker heartbeats with multiple block masters.
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java:41]: done

File: core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java [hunk 34]
```diff
@@ -0,0 +1,179 @@
...
+    }
+  }
+
+  /**
+   * Starts the heartbeats.
+   * @param executorService the executor service to run the heartbeats
+   */
+  public synchronized void start(ExecutorService executorService) {
+    if (!mStarted) {
+      mStarted = true;
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/BlockSyncMasterGroup.java:88]: i also feel probably no, but i just added it for safety.

File: core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java [hunk 35]
```diff
@@ -138,7 +138,8 @@ public class DefaultBlockWorker extends AbstractWorker implements BlockWorker {
   public DefaultBlockWorker(BlockMasterClientPool blockMasterClientPool,
       FileSystemMasterClient fileSystemMasterClient, Sessions sessions, BlockStore blockStore,
       AtomicReference<Long> workerId) {
-    super(ExecutorServiceFactories.fixedThreadPool("block-worker-executor", 5));
+    super(ExecutorServiceFactories.fixedThreadPool("block-worker-executor",
+        Configuration.getBoolean(PropertyKey.WORKER_REGISTER_TO_ALL_MASTERS) ? 10 : 5));
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/DefaultBlockWorker.java:142]: yeah i feel cached thread pool is better. there isn't this many threads anyways

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java [hunk 36]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * The block master sync thread when workers are registered to all masters.
+ * With respect to behaviors, this implementation differs from {@link BlockMasterSync} in:
+ * 1. The registration takes place asynchronously and the caller can poll the registration state.
+ *  We need to make the process async because when standby master read is enabled, workers have to
+ *  register to all masters and these registrations can happen concurrently to speed up the process.
+ * 2. A registration fail doesn't throw a fatal exception. Instead, it retries endlessly.
+ *  This is because a standby master registration failure
+ *  should not a soft failure and can be retried later.
```
[the remark this answers is not in the document]
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:46]: done

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java (hunk 26 above)
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:156]: worker oom 
because worker need to get the block location list from the metastore for each master when it registers with the masters. If these registration happen at the same time, these block location list info might cause worker OOM

Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:156]: I updated the comment

File: core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java (hunk 28 above)
Comment (elega) [reply core/server/worker/src/main/java/alluxio/worker/block/SpecificMasterBlockSync.java:263]: yeah this is a good catch. i updated DefaultBlockMaster and made the standby master never return a Free command

File: core/common/src/main/java/alluxio/metrics/MetricKey.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (elega) [reply core/common/src/main/java/alluxio/metrics/MetricKey.java:2065]: okay so we discussed offline and the conclusion is that we merge this metrics for now and we can iterate if users need more visibility or clarity on this

**Review by jiacheliu3 — APPROVED** [review]
LGTM, thanks for the work!

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please
